### ISSUE TITLE
exclude(project(...)) doesn't exlude transitive dependencies

### ISSUE THREAD

**Issue by DmitriyMV** [GradleUp/shadow#64]
As for 1.0.1 all dependencies that are defined in sub-project, but not in main project will still go into uber jar. 0.9.0-M3 had different(working) behavior. No way to completely exclude sub project dependencies because:
    exclude(dependency(...))
doesn't exclude transitive dependencies.

**Comment by rocketraman** [comment]
BTW, for anyone that comes here via Google looking to exclude a particular jar _and all its transitive dependencies_ (this is useful for creating jars that can be executed on storm or flink, for example), here is another approach that seems to work well:

http://unethicalblogger.com/2015/07/15/gradle-goodness-excluding-depends-from-shadow.html

TL;DR: exclude the jar from the runtime configuration e.g.

```
configurations {
  runtime.exclude group: 'org.apache.flink'
}
```

In the case of storm/flink, rather than a hack, this does reflect reality in a way: these libs will already be available at runtime so don't need to be included in the "runtime" configuration.

**Comment by johnrengelman** [comment]
a better solution is to use the new builtin `compileOnly` configuration in Gradle 3.x. Those libraries will not be merged into the shadow jar.

**Comment by rocketraman** [comment]
> a better solution is to use the new builtin compileOnly configuration in Gradle 3.x

Nice, thank you. This would be useful in the documentation.

**Comment by rocketraman** [comment]
> a better solution is to use the new builtin compileOnly configuration in Gradle 3.x. Those libraries will not be merged into the shadow jar. compileOnly is effectively Maven's provided scope.

Hmm, I just tried this with Gradle 3.1 and plugin 1.2.3 and it didn't work... does this require an as yet unreleased version of the plugin?

**Comment by johnrengelman** [comment]
it shouldn't. I thought i tested this...what doesn't work with it? What version of Gradle are you using?

**Comment by johnrengelman** [comment]
I added a test for this early this year: https://github.com/johnrengelman/shadow/commit/ecc8168c34f5419f5d26fe489274c0745615a4e1

**Comment by rocketraman** [comment]
> it shouldn't. I thought i tested this...what doesn't work with it? What version of Gradle are you using?

Gradle 3.1, shadow plugin 1.2.3. With `compileOnly`, the debug stats are:

```
16:54:13.039 [INFO] [org.gradle.api.Task] *******************
GRADLE SHADOW STATS

Total Jars: 90 (includes project)
Total Time: 6.817s [6817ms]
Average Time/Jar: 0.0757444444444s [75.7444444444ms]
```

With the `runtime.exclude` approach with the same jars:

```
16:56:02.996 [INFO] [org.gradle.api.Task] *******************
GRADLE SHADOW STATS

Total Jars: 22 (includes project)
Total Time: 0.616s [616ms]
Average Time/Jar: 0.028s [28.0ms]
*******************
```

**Comment by johnrengelman** [comment]
can you post the 2 gradle files that you are using?

**Comment by rocketraman** [comment]
Ah I see what the issue is... I have the following dependencies:

```
dependencies {
  compileOnly "org.apache.flink:flink-java:1.1.3"
  compileOnly "org.apache.flink:flink-streaming-java_2.11:1.1.3"
  compileOnly "org.apache.flink:flink-clients_2.11:1.1.3"
  compile "org.apache.flink:flink-connector-wikiedits_2.11:1.1.3"
  compile 'org.apache.flink:flink-connector-kafka-0.9_2.11:1.1.3'
}
```

I need to include the connectors because they are not part of the flink runtime. However, the connectors have transitive compile dependencies on `flink-streaming-java` and others, which adds them back to the `compile` and thus `runtime` configurations.

So I guess in this case, excluding via `runtime.exclude` is indeed the better approach?

**Comment by johnrengelman** [comment]
Yeah, that gets a bit weird. I would argue that the connector libraries are publishing invalid POMs (the POMs should be setting the `flink-clients` and `flink-streaming-java` libraries as `provided` scope instead of `compile`.
So you could also do this:
```
dependencies {
  compileOnly "org.apache.flink:flink-java:1.1.3"
  compileOnly "org.apache.flink:flink-streaming-java_2.11:1.1.3"
  compileOnly "org.apache.flink:flink-clients_2.11:1.1.3"
  compile("org.apache.flink:flink-connector-wikiedits_2.11:1.1.3") {
    exclude group: 'org.apache.flink', module: 'flink-clients_2.11'
    exclude group: 'org.apache.flink', module: 'flink-streaming-java_2.11'
  }
  compile('org.apache.flink:flink-connector-kafka-0.9_2.11:1.1.3') {
    exclude group: 'org.apache.flink', module: 'flink-clients_2.11'
    exclude group: 'org.apache.flink', module: 'flink-streaming-java_2.11'
  }
}
```

**Comment by rocketraman** [comment]
> I would argue that the connector libraries are publishing invalid POMs (the POMs should be setting the flink-clients and flink-streaming-java libraries as provided scope instead of compile.

Agreed.

> So you could also do this:

Yup, this works also.

I noticed you have issue #257 for improving docs related to this -- some of the lessons learned above would be useful there too.

**Comment by johnrengelman** [comment]
Yeah, over the last week or so, i've been thinking about the usefulness of `configurations.shadow`. Not sure it's entirely needed anymore with `compileOnly` and it's behavior can add some often surprising side-effects. (i.e. `configurations.shadow` become `runtime`  scope for the shadow jar artifact). Definitely will be adding better docs around this soon.

**Comment by bucknerns** [comment]
If anyone ever runs into this thread after spending hours messing with this I ended up following this https://kennethjorgensen.com/blog/2014/fat-jars-with-excluded-dependencies-in-gradle
but replaced the jar task and configurations with this

    configurations {
        provided
        implementation.extendsFrom provided //implementation is useful for compiling with spark dependencies but having a smaller jar since they are already in the classpath
    }

    shadowJar {
        zip64 = true
        mergeServiceFiles()
        // only if you need to relocate a package
        // With aws emr you will need to relocate protobuf 3 libs
        relocate('com.google.protobuf', 'shaded.com.google.protobuf') 
        archiveFileName = "originaljarname.jar" // If you don't want the -all added to the jar
    
        def exclude_modules = project
                .configurations
                .provided
                .resolvedConfiguration
                .getLenientConfiguration()
                .getAllModuleDependencies()
                .collect {
                    it.name
                }
        dependencies {
            exclude(dependency {
                exclude_modules.contains(it.name)
            })
        }
    }
    
    tasks.jar.enabled = false // disable the original jar so I don't build 2
    tasks.assemble.dependsOn tasks.shadowJar // make shadowjar automatically run with assemble/build is called

**Comment by brainbytes42** [comment]
> a better solution is to use the new builtin `compileOnly` configuration in Gradle 3.x. Those libraries will not be merged into the shadow jar. `compileOnly` is effectively Maven's `provided` scope. So you would do this:
> 
> ```
> dependencies {
>   compileOnly "org.apache.flink:flink:version"
> }
> ```

Should this approach work for transitive dependencies also? (For me, it doesn't, unfortunately.)

With Gradle 8.6 / Shadow 8.1.1, compileOnly suppresses the specific reference from being included in the jar, but all its dependencies are included.

For some plugins, I have the only dependency on the framework, which provides it's dependencies at runtime. So I'd like to have the framework and all its dependecies excluded in the plugin's jar... (But any additional library needed only inside the plugin needs to be shadowed, otherwise, I could just use normal jar...! ;-) )

```
dependencies {
    compileOnly project(':MyFramework')
}
```

With exclude(), I need around 20 lines to exclude all dependecies by hand - and I'm sure, Ill forget to adapt this, if I update or add dependencies to the framework... :-/
